tree_in_context/src/utility/in_context.egg

```diff
@@ -1,146 +1,219 @@
+;; This is because a loop (DoWhile inputs outputs)
+;; can become equal to
+;; (DoWhile inputs ctx1)
+;; where ctx1 = (InContext (InLoop inputs outputs) outputs)
+;; Which we can add context to again:
+;; (DoWhile inputs (InContext (InLoop inputs ctx1) ctx1))
+;; These two contexts are equal, since the set of reachable
+;; values is the same.
+(rule ((= body_with_context
+          (DoAddContext seen  (InLoop inputs outputs) (Full) outputs)) ;; a loop body with context
+       (DoAddContext    seen2 (InLoop inputs body_with_context) ;; same inputs
+                    (Full) body_with_context) ;; a loop body whose context already has context
+       )
+      ((union (InLoop inputs outputs) (InLoop inputs body_with_context)))
+      :ruleset context-helpers)
+
+;; key rule that detects cycles
+;; It finds that we have already seen the eclass we are adding context to.
+;; The corresponding eclasses are unioned, and the cyclic
+;; substitution is subsumed to ensure saturation.
+(rule ((= lhs (DoAddContext seen ctx scope in))
+       (Path-contains seen
+         (PathCons in original rest)))
+      ((union lhs original)
+       (subsume (DoAddContext seen ctx scope in)))
+      :ruleset context-helpers)
+
+
+
+
+;; ############################## Base cases- leaf nodes
+
+(rewrite (DoAddContext seen ctx scope (Arg ty))
+         (InContext ctx (Arg ty))
+         :ruleset context-helpers)
+
+(rewrite (DoAddContext seen ctx scope (Const constant ty))
+         (InContext ctx (Const constant ty))
+         :ruleset context-helpers)
+
+(rewrite (DoAddContext seen ctx scope (Empty ty))
+         (InContext ctx (Empty ty))
+         :ruleset context-helpers)
+
+
+;; ######################################### Operators
+
+
+(rewrite (DoAddContext seen ctx scope (Bop op c1 c2))
+         (Bop op
+           (DoAddContext seen ctx scope c1)
+           (DoAddContext seen ctx scope c2))
+               :ruleset context-helpers)
+(rewrite (DoAddContext seen ctx scope (Uop op c1))
+         (Uop op (DoAddContext seen ctx scope c1))
+         :ruleset context-helpers)
+(rewrite (DoAddContext seen ctx scope (Get c1 index))
+         (Get (DoAddContext seen ctx scope c1) index)
+               :ruleset context-helpers)
+(rewrite (DoAddContext seen ctx scope (Alloc c1 ty))
+         (Alloc (DoAddContext seen ctx scope c1) ty)
+         :ruleset context-helpers)
+(rewrite (DoAddContext seen ctx scope (Call name c1))
+         (Call name (DoAddContext seen ctx scope c1))
+         :ruleset context-helpers)
+
+(rewrite (DoAddContext seen ctx scope (Single c1))
+         (Single (DoAddContext seen ctx scope c1))
+         :ruleset context-helpers)
+(rewrite (DoAddContext seen ctx scope (Concat order c1 c2))
+         (Concat order
+           (DoAddContext seen ctx scope c1)
+           (DoAddContext seen ctx scope c2))
+         :ruleset context-helpers)
+
+;; Replace existing contexts.
+;; This relies on the fact that new contexts must be more specific than old ones.
```

Comment:
I know we discussed this in a meeting at some point, but have we written down why this is true anywhere (and continues to be true, say, when we add some of these contexts ahead of time in the Rust pass?)

Reply:
That's a great question, and I hope to answer that soon by updating the semantics in the repository.
A high-level answer is that we maintain this invariant:
A context node defines the set of values that a program is valid on, and this program is *guaranteed to execute* on these values. That means nested context nodes must pass for all values the program is executed on.